core/src/main/java/io/questdb/griffin/model/IntervalUtils.java

```diff
@@ -366,103 +398,477 @@ public static long parseFloorPartialDate(CharSequence seq) throws NumericExcepti
+                }
+            }
+            throw SqlException.$(position, "Invalid date");
+        }
+    }
+
+    public static void subtract(LongList intervals, int divider) {
+        IntervalUtils.invert(intervals, divider);
+        IntervalUtils.intersectInplace(intervals, divider);
+    }
+
+    public static long tryParseTimestamp(CharSequence seq) throws CairoException {
+        try {
+            return parseFloorPartialDate(seq, 0, seq.length());
+        } catch (NumericException e) {
+            throw CairoException.instance(0).put("Invalid timestamp: ").put(seq);
+        }
+    }
+
+    private static int tenPow(int i) throws NumericException {
+        switch (i) {
+            case 0:
+                return 1;
+            case 1:
+                return 10;
+            case 2:
+                return 100;
+            case 3:
+                return 1000;
+            case 4:
+                return 10000;
+            case 5:
+                return 100000;
+            default:
+                throw NumericException.INSTANCE;
+        }
+    }
+
+    private static long checkTimezoneTail(CharSequence seq, int p, int lim) throws NumericException {
+        if (lim == p) {
+            return 0;
+        }
+
+        if (lim - p < 2) {
+            checkChar(seq, p, lim, 'Z');
+            return 0;
+        }
+
+        if (lim - p > 2) {
+            int tzSign = parseSign(seq, p++);
+            int hour = Numbers.parseInt(seq, p, p += 2);
+            checkRange(hour, 0, 23);
+
+            if (lim - p == 3) {
+                // Optional : separator between hours and mins in timezone
+                checkChar(seq, p++, lim, ':');
+            }
+
+            if (checkLenStrict(p, lim)) {
+                int min = Numbers.parseInt(seq, p, p + 2);
+                checkRange(min, 0, 59);
+                return tzSign * (hour * Timestamps.HOUR_MICROS + min * Timestamps.MINUTE_MICROS);
+            } else {
+                return tzSign * (hour * Timestamps.HOUR_MICROS);
+            }
+        }
+        throw NumericException.INSTANCE;
+    }
+
+    private static int parseSign(CharSequence seq, int p) throws NumericException {
+        int tzSign;
+        switch (seq.charAt(p)) {
+            case '+':
+                tzSign = 1;
+                break;
+            case '-':
+                tzSign = -1;
+                break;
+            default:
+                throw NumericException.INSTANCE;
+        }
+        return tzSign;
+    }
+
+    /**
+     * Unions two lists of intervals compacted in one list in place.
+     * Intervals to be chronologically ordered and result list will be ordered as well.
+     * <p>
+     * Treat a as 2 lists,
+     * a: first from 0 to divider
+     * b: from divider to the end of list
+     *
+     * @param intervals 2 lists of intervals concatenated in 1
+     */
+    static void unionInplace(LongList intervals, int dividerIndex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.